src/util/metricsAggregator.js

```diff
@@ -0,0 +1,167 @@
+/* eslint-disable */
+const cluster = require('cluster');
+const logger = require('../logger');
+const { Worker, isMainThread } = require('worker_threads');
+const GET_METRICS_REQ = 'rudder-transformer:getMetricsReq';
+const GET_METRICS_RES = 'rudder-transformer:getMetricsRes';
+const AGGREGATE_METRICS_REQ = 'rudder-transformer:aggregateMetricsReq';
+const AGGREGATE_METRICS_RES = 'rudder-transformer:aggregateMetricsRes';
+const RESET_METRICS_REQUEST = 'rudder-transformer:resetMetricsReq';
+const METRICS_AGGREGATOR_PERIODIC_RESET_ENABLED =
+  process.env.METRICS_AGGREGATOR_PERIODIC_RESET_ENABLED === 'true';
+const METRICS_AGGREGATOR_PERIODIC_RESET_INTERVAL_SECONDS = process.env
+  .METRICS_AGGREGATOR_PERIODIC_RESET_INTERVAL_SECONDS
+  ? parseInt(process.env.METRICS_AGGREGATOR_PERIODIC_RESET_INTERVAL_SECONDS, 10)
+  : 30 * 60;
+
+class MetricsAggregator {
+  constructor(prometheusInstance) {
+    this.metricsBuffer = [];
+    this.pendingMetricRequests = 0;
+    this.resolveFunc = null;
+    this.rejectFunc = null;
+    this.prometheusInstance = prometheusInstance;
+    this.createWorkerThread();
+    this.registerCallbacks();
+  }
+
+  registerCallbacks() {
+    if (cluster.isPrimary) {
+      // register callback for master process
+      cluster.on('message', async (worker, message) => {
+        if (message.type === GET_METRICS_RES) {
+          logger.debug(`[MetricsAggregator] Master received metrics from worker ${worker.id}`);
+          await this.handleMetricsResponse(message);
+        }
+      });
+      // register callback to reset metrics if enabled
+      if (METRICS_AGGREGATOR_PERIODIC_RESET_ENABLED) {
+        this.registerCallbackForPeriodicReset(METRICS_AGGREGATOR_PERIODIC_RESET_INTERVAL_SECONDS);
+      }
+      return;
+    }
+    // register callback for worker process
+    cluster.worker.on('message', async (message) => {
+      if (message.type === GET_METRICS_REQ) {
+        logger.debug(`[MetricsAggregator] Worker ${cluster.worker.id} received metrics request`);
+        try {
+          const metrics = await this.prometheusInstance.prometheusRegistry.getMetricsAsJSON();
+          cluster.worker.send({ type: GET_METRICS_RES, metrics });
+        } catch (error) {
+          cluster.worker.send({ type: GET_METRICS_RES, error: error.message });
+        }
+      } else if (message.type === RESET_METRICS_REQUEST) {
+        logger.info(
+          `[MetricsAggregator] Worker ${cluster.worker.id} received reset metrics request`,
+        );
+        this.prometheusInstance.prometheusRegistry.resetMetrics();
+        logger.info(`[MetricsAggregator] Worker ${cluster.worker.id} reset metrics successfully`);
+      }
+    });
+  }
+
+  registerCallbackForPeriodicReset(intervalSeconds) {
+    // store the timer in the aggregator for future operations like shutdown
+    this.periodicResetTimer = setInterval(() => {
```

Comment:
probably nitpicking for this context - but it might be better to use setTimeout and register another setTimeout after every reset. setInterval does not check if previous operation completed or not

Reply:
in our usecase, the resetMetrics function called here only sends messages to workers asking them to reset metrics and returns 
will keep this in mind going forward when using setInterval